base/compiler/ssair/passes.jl

```diff
@@ -921,15 +932,68 @@ function sroa_pass!(ir::IRCode)
         used_ssas = copy(compact.used_ssas)
         simple_dce!(compact, (x::SSAValue) -> used_ssas[x.id] -= 1)
         ir = complete(compact)
-        sroa_mutables!(ir, defuses, used_ssas, lazydomtree)
+        sroa_mutables!(ir, defuses, used_ssas, lazydomtree, inlining)
         return ir
     else
         simple_dce!(compact)
         return complete(compact)
     end
 end
 
-function sroa_mutables!(ir::IRCode, defuses::IdDict{Int, Tuple{SPCSet, SSADefUse}}, used_ssas::Vector{Int}, lazydomtree::LazyDomtree)
+function try_inline_finalizer!(ir::IRCode, argexprs::Vector{Any}, idx::Int, mi::MethodInstance, inlining::InliningState)
+    code = get(inlining.mi_cache, mi, nothing)
+    if code isa CodeInstance
+        if use_const_api(code)
+            # No code in the function - Nothing to do
+            inlining.et !== nothing && push!(inlining.et, mi)
+            return true
+        end
+        src = code.inferred
+    else
+        src = code
+    end
+
+    src = inlining_policy(inlining.interp, src, IR_FLAG_NULL, mi, Any[])
+    src === nothing && return false
+    src = retrieve_ir_for_inlining(mi, src)
+
+    # For now: Require finalizer to only have one basic block
+    length(src.cfg.blocks) == 1 || return false
+
+    # Ok, we're committed to inlining the finalizer
+    inlining.et !== nothing && push!(inlining.et, mi)
+
+    linetable_offset, extra_coverage_line = ir_inline_linetable!(ir.linetable, src, mi.def, ir[SSAValue(idx)][:line])
+    if extra_coverage_line != 0
+        insert_node!(ir, idx, NewInstruction(Expr(:code_coverage_effect), Nothing, extra_coverage_line))
+    end
+
+    # TODO: Use the actual inliner here rather than open coding this special
+    # purpose inliner.
+    spvals = mi.sparam_vals
+    ssa_rename = Vector{Any}(undef, length(src.stmts))
+    for idx′ = 1:length(src.stmts)
+        urs = userefs(src[SSAValue(idx′)][:inst])
+        for ur in urs
+            if isa(ur[], SSAValue)
+                ur[] = ssa_rename[ur[].id]
+            elseif isa(ur[], Argument)
+                ur[] = argexprs[ur[].n]
+            elseif isexpr(ur[], :static_parameter)
+                ur[] = spvals[ur[].args[1]]
+            end
+        end
+        # TODO: Scan newly added statement into the sroa defuse struct
```

Comment:
Is this something for Ian? I thought he just merged the PR that fixed this for the other cases here so that he now relies on this.

Reply:
This is not compact, so it doesn't have a correctness impact. What this TODO refers to is the ability to do a mutable SROA on the newly inlined code, which essentially requires running the initial loop of the sroa pass on the newly inserted instructions.